game/plugin/skills/woodcutting/src/woodcutting.plugin.kts

```diff
@@ -0,0 +1,122 @@
+import org.apollo.game.model.entity.Skill
+import org.apollo.game.model.entity.obj.GameObject
+import org.apollo.game.plugin.skills.woodcutting.*
+import org.apollo.game.plugins.api.woodcutting
+import org.apollo.game.plugins.api.skills
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+class WoodcuttingTarget(val objectId: Int, val position: Position, val tree: Tree) {
+
+    /**
+     * Get the tree object in the world
+     */
+    fun getObject(world: World): Optional<GameObject> {
+        val region = world.regionRepository.fromPosition(position)
+        return region.findObject(position, objectId)
+    }
+
+    /**
+     * Check if the tree was cut down
+     */
+    fun isCutDown(mob: Player): Boolean {
+        return rand(100) <= tree.chance * 100
+    }
+}
+
+class WoodcuttingAction(val player: Player, val tool: Axe, val target: WoodcuttingTarget) : AsyncDistancedAction<Player>(PULSES, true, player, target.position, TREE_SIZE) {
+
+    companion object {
+        private val PULSES = 0
+        private val TREE_SIZE = 2
+        private val MINIMUM_RESPAWN_TIME = 30L //In seconds
+
+        /**
+         * Find the highest level axe the player has
+         */
+        fun axeFor(player: Player): Axe? {
+            return getAxes()
+                    .filter { it.level <= player.skills.woodcutting.currentLevel }
+                    .filter { player.equipment.contains(it.id) || player.inventory.contains(it.id) }
+                    .sortedByDescending { it.level }
+                    .firstOrNull()
+        }
+
+        /**
+         * Starts a [WoodcuttingAction] for the specified [Player], terminating the [Message] that triggered it.
+         */
+        fun start(message: ObjectActionMessage, player: Player, wood: Tree) {
+            val axe = axeFor(player)
+            if (axe != null) {
+                val action = WoodcuttingAction(player, axe, WoodcuttingTarget(message.id, message.position, wood))
+                player.startAction(action)
+            } else {
+                player.sendMessage("You do not have a axe for which you have the level to use.")
+            }
+
+            message.terminate()
+        }
+    }
+
+    override fun start() {
+        mob.turnTo(position)
+
+        val level = mob.skills.woodcutting.currentLevel
+
+        if (level < target.tree.level) {
+            mob.sendMessage("You do not have the required level to cut down this tree.")
+            stop()
+        }
+    }
+
+    override suspend fun executeActionAsync() {
+        mob.sendMessage("You swing your axe at the tree.")
+        mob.playAnimation(tool.animation)
+
+        wait(tool.pulses)
+
+        //Check that the object exists in the world
+        val obj = target.getObject(mob.world)
+        if (!obj.isPresent) {
+            stop()
+            return
+        }
+
+        //Check that the player had room in their inventory
+        if (mob.inventory.freeSlots() == 0) {
```

Comment:
Apollo already handles capacity checking and notification (i.e. `forceCapacityExceeded()`) when you attempt to add items. However, in RS it actually checks if you have inventory space at the start of the action (before you play the animation), so basically we want to move this check up to the very top of this function.

Reply:
Good to know. I will move this to the start in the companion object.